### ISSUE TITLE
Travis build failing on macOS Sierra (10.12)

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/spinalcordtoolbox#2841]
https://travis-ci.org/github/neuropoly/spinalcordtoolbox/jobs/717990889

should we drop support?

**Comment by joshuacwnewton** [comment]
It looks like this is less about 10.12 and more about how we install `scikit-image`. 

For 10.13/10.14, [`scikit-image` provides a wheel to install](https://pypi.org/project/scikit-image/#files), but for 10.12 we have to install the sdist. For the sdist, it looks like we're running into a similar issue to https://github.com/scikit-learn/scikit-learn/issues/4164. 

The tl;dr is that `setup.py` in the `scikit-image` sdist requires `numpy` to be installed beforehand (i.e. a sequential install). But, `./install_sct` uses `pip install -r requirements.txt` for everything, so packages aren't installed sequentially.

I think this could be fixed by changing this line in our install script such that numpy is installed first.

https://github.com/neuropoly/spinalcordtoolbox/blob/b64cad3c846fd6bd7a557688b67b80fe0b2c6dc2/install_sct#L539

**Comment by jcohenadad** [comment]
> I think this could be fixed by changing this line in our install script such that numpy is installed first.

great! let's do it 👍

**Comment by joshuacwnewton** [comment]
Oops, looks like the same conclusion was arrived at for #2840. @kousu and @Drulex figured this out hours ago. :P

**Comment by kousu** [comment]
Getting rid of the explicit numpy was something I proposed (#2750). There was a good reason: installing numpy in requirements.txt and in install_sct made pip install numpy twice, slowing down installation (and possibly causing inconsistencies if we version one and forget to version the other); @Drulex did some snappy research and pointed us at this old thread https://github.com/scipy/scipy/issues/3379

> Not an issue we want to fix. At least until the ridiculous --upgrade behavior of setuptools is killed.
> ...
> The problem about declaring the dependency is that it very frequently involves a lengthy recompile and install of numpy when there is already an adequate numpy present (that setuptools does not accept for whatever reason).
> ...
> Yes. It's not that hard with only numpy as dependency, and certainly better than introducing issues for users due to pip/setuptools trying to upgrade numpy when that's not what they asked for.
> ...
> The problem really lies with the poor quality Python build/install tools

and their workaround code https://github.com/scipy/scipy/blob/834f756ebb4eb04049ed7bc26226a27fb58ca5ad/setup.py#L190 which I'm going to quote for posterity

```
    # Figure out whether to add ``*_requires = ['numpy']``.
    # We don't want to do that unconditionally, because we risk updating
    # an installed numpy which fails too often.  Just if it's not installed, we
    # may give it a try.  See gh-3379.
    build_requires = []
    try:
        import numpy
    except:
        build_requires = ['numpy>=1.5.1']
```

but that was 6 years ago. Surely there's a better solution now? Maybe not? ...maybe not, here's the latest: https://github.com/scipy/scipy/blob/2ed8169c5c4c9181cd38a52db614c692eb802c15/setup.py#L508-L521

```
    # Figure out whether to add ``*_requires = ['numpy']``.
    # We don't want to do that unconditionally, because we risk updating
    # an installed numpy which fails too often.  Just if it's not installed, we
    # may give it a try.  See gh-3379.
    try:
        import numpy
    except ImportError:  # We do not have numpy installed
        build_requires = ['numpy>=1.14.5']
    else:
        # If we're building a wheel, assume there already exist numpy wheels
        # for this platform, so it is safe to add numpy to build requirements.
        # See gh-5184.
        build_requires = (['numpy>=1.14.5'] if 'bdist_wheel' in sys.argv[1:]
                          else [])
```

I tried installing scipy from source (in a fresh venv):

```
(.venv36) [kousu@laptop scipy]$ python setup.py install

Note: if you need reliable uninstall behavior, then install
with pip instead of using `setup.py install`:

  - `pip install .`       (from a git repo or downloaded source
                           release)
  - `pip install scipy`   (last SciPy release on PyPI)


/home/kousu/src/neuropoly/.venv36/lib/python3.6/site-packages/setuptools/distutils_patch.py:26: UserWarning: Distutils was imported before Setuptools. This usage is discouraged and may exhibit undesirable behaviors or errors. Please use Setuptools' objects directly or at least import Setuptools first.
  "Distutils was imported before Setuptools. This usage is discouraged "
Traceback (most recent call last):
  File "setup.py", line 590, in <module>
    setup_package()
  File "setup.py", line 565, in setup_package
    from numpy.distutils.core import setup
ModuleNotFoundError: No module named 'numpy'
```

so it looks like their `build_requires` hack quoted above is just dead in the water. It's not working any more.

However, `pip` works:

```
(.venv36) [kousu@laptop scipy]$ pip install .
Processing /home/kousu/src/neuropoly/scipy
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
```

this is because they're using the new pyproject.toml cruft:

```
$ cat pyproject.toml 
[build-system]
requires = [
    "wheel",
    "setuptools",
    "Cython>=0.29.18",
    "numpy==1.14.5; python_version=='3.6' and platform_system!='AIX'",
    "numpy==1.14.5; python_version=='3.7' and platform_system!='AIX'",
    "numpy==1.17.3; python_version>='3.8' and platform_system!='AIX'",
    "numpy==1.16.0; python_version=='3.6' and platform_system=='AIX'",
    "numpy==1.16.0; python_version=='3.7' and platform_system=='AIX'",
    "numpy==1.17.3; python_version>='3.8' and platform_system=='AIX'",
    "pybind11>=2.4.3",
]
```

--


Also, there's a comment in our requirements.txt that numpy is on top *on purpose* but it looks like at some point since then `pip` has changed and now installs...in alphabetical order? One of the comments that @Drulex found explicitly re-enables ordering (https://github.com/scikit-learn/scikit-learn/issues/4164#issuecomment-243259447). So that workaround of ours isn't good anymore.

--

I think I can fix this. The problem is that scikit-image isn't using pyproject.toml, which is what modern pip requires. And this wasn't caught because they install build deps with `pip install -r`: 

https://github.com/scikit-image/scikit-image/blob/59fd9682aad41a1f69a35f05d2a7e15ebdfb8b29/tools/travis/before_install.sh#L78

So I think this is a bug in scikit-image. They're *not* installable from source, at least not the way pip expects (which I think should count as the standard), and that is breaking us because they're not producing Sierra wheels.

**Comment by kousu** [comment]
This is such a wild rabbit hole.

**Comment by joshuacwnewton** [comment]
@kousu Not to distract from the rabbit hole, but, as far as a practical short-term solution, how awful would something [like this](https://stackoverflow.com/a/5395951) be? Especially considering our install script is already a shell script.

**Comment by kousu** [comment]
mmm that's alright, that's the same fix in https://github.com/scikit-learn/scikit-learn/issues/4164#issuecomment-243259447, but it's fighting against the tide of history (or the tide of pypa at least).

Since the issue is just about numpy, and it's because numpy is a secret build dependency, if we were to fix it in `./install_sct` I would:

1. put back `pip install numpy` (removed, by my hand, in #2751 :( )
2. remove numpy from requirements.txt

If I can't get my hacks in #2840 to work I'll just do that.

**Comment by kousu** [comment]
And yes I believe it does do fancy resolving.

**Comment by joshuacwnewton** [comment]
> 1. put back pip install numpy (removed, by my hand, in #2751 :( )
> 2. remove numpy from requirements.txt

Is `requirements.txt` used anywhere else? e.g. pip installing for RTD? I can't recall if it's necessary for it to have everything we need.

**Comment by kousu** [comment]
It is. It's the only place we specify dependencies (and I think we have to change that https://github.com/neuropoly/spinalcordtoolbox/issues/1526#issuecomment-674330466, eventually)

scikit-image goes all-in on requirements.txt. Look, they have a separate one for each use case:

* https://github.com/scikit-image/scikit-image/blob/59fd9682aad41a1f69a35f05d2a7e15ebdfb8b29/requirements/build.txt
* https://github.com/scikit-image/scikit-image/blob/59fd9682aad41a1f69a35f05d2a7e15ebdfb8b29/requirements/default.txt
* https://github.com/scikit-image/scikit-image/blob/59fd9682aad41a1f69a35f05d2a7e15ebdfb8b29/requirements/test.txt
* https://github.com/scikit-image/scikit-image/blob/59fd9682aad41a1f69a35f05d2a7e15ebdfb8b29/requirements/docs.txt

(something covered by [`extras_requires`](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies); we've got it in our own setup.py in fact, but we don't have `install_requires`, strangely).

but installing like this is not compatible with `pip install scikit-image`. It just isn't. That's not how pip works anymore, if it ever did. The only way `pip install scikit-image` works is because they're doing `pip install -r requirements/build.txt; pip setup.py build;` across all the platforms and uploading the results to pypi so that everyone can download precompiled copies.

**Comment by kousu** [comment]
Okay I've fixed it: https://github.com/scikit-image/scikit-image/pull/4920 :rocket: but it will probably be a while before that makes it out to pypi, so we should still apply the workaround in #2840.

**Comment by hmaarrfk** [comment]
> > I think this could be fixed by changing this line in our install script such that numpy is installed first.
> 
> great! let's do it 👍

I suggest you follow this route. 

Getting bleeding edge python to work correctly before it is released is an uphill battle. Now that python 3 vs python 2 has settled, it is likely that the migration toward 3.9 will be rather smooth once dependencies are satisfied. 

While it may be possible to fix things for python 3.9, your strategy will never be tested again until 3.10 gets released, at which point there may be a totally different issue.

**Comment by hmaarrfk** [comment]
PS. I'm a core dev of scikit image, I can be swayed to incorporating your patch (even though I was the one who pushed to remove pyproject toml) but I want to help you get what you need accomplished in your timeline.

**Comment by kousu** [comment]
> PS. I'm a core dev of scikit image, I can be swayed to incorporating your patch (even though I was the one who pushed to remove pyproject toml) but I want to help you get what you need accomplished in your timeline.

Oh interesting. Is there a thread about why you avoided `pyproject.toml`?

fwiw, pyproject.toml seems really unstable to me, as it stands, and I would avoid it too. But `pyproject.toml` is what pip wants to use now to accomplish what `setup_requires` was supposed to (but..never quite did?)

**Comment by joshuacwnewton** [comment]
Regardless of the eventual conclusion, thank you greatly @hmaarrfk for taking the time to investigate and offer solutions. :smiling_face_with_three_hearts:

**Comment by kousu** [comment]
> While it may be possible to fix things for python 3.9, your strategy will never be tested again until 3.10 gets released, at which point there may be a totally different issue.

From my perspective, this isn't about getting it working on cpython3.9, it's about having a source-clean build environment. 
As it stands, `pip install spinalcordtoolbox` requires you to be on an OS where [you've released binaries](https://github.com/scikit-image/scikit-image-wheels/). To go from source code to a working project requires `pip install numpy` as an extra pre-step.

**Comment by hmaarrfk** [comment]
> From my perspective, this isn't about getting it working on cpython3.9, it's about having a source-clean build environment.
As it stands, pip install spinalcordtoolbox requires you to be on an OS where you've released binaries. To go from source code to a working project requires pip install numpy as an extra pre-step.

This is the age old problem with relying on dependencies.

The problem with allowing users to build on unsupported platforms is that while thye may be able to install things, it is totally unclear if it will work as intended, or what bugs might occur.

I am very interested in adding support for more platforms.

> fwiw, pyproject.toml seems really unstable to me, as it stands, and I would avoid it too.

This is exactly why I pushed to remove it.

I would encourage you to help us build the infrastructure to support the platforms you need.

I think grlee77 has also made a few comments that would bring the the PR up to par.
